pkg/virtual/framework/transforming/transformers.go

```diff
@@ -0,0 +1,229 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package transforming
+
+import (
+	"context"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/klog/v2"
+)
+
+type UpdateSelectorsFunc func(ctx context.Context, labelSelector labels.Selector, fieldSelector fields.Selector) (labels.Selector, fields.Selector, error)
+
+func TransformsSelectors(transformerName string, updateFunc UpdateSelectorsFunc) Transformer {
+
+	updateSelectorString := func(ctx context.Context, labelSelectorStr string, fieldSelectorStr string) (newLabelSelectorStr string, newFieldSelectorStr string, err error) {
+		labelSelector, err := labels.Parse(labelSelectorStr)
+		if err != nil {
+			return "", "", err
+		}
+		fieldSelector, err := fields.ParseSelector(fieldSelectorStr)
+		if err != nil {
+			return "", "", err
+		}
+		newLabelSelector, newFieldSelector, err := updateFunc(ctx, labelSelector, fieldSelector)
+		if err != nil {
+			return "", "", err
+		}
+		newLabelSelectorStr = newLabelSelector.String()
+		newFieldSelectorStr = newFieldSelector.String()
+		return
+	}
+
+	return Transformer{
+		Name: transformerName,
+		BeforeList: func(ctx context.Context, opts metav1.ListOptions) (context.Context, metav1.ListOptions, error) {
+			newLabelSelector, newFieldSelector, err := updateSelectorString(ctx, opts.LabelSelector, opts.FieldSelector)
+			if err != nil {
+				return ctx, opts, err
+			}
+			opts.LabelSelector = newLabelSelector
+			opts.FieldSelector = newFieldSelector
+			return ctx, opts, nil
+		},
+		BeforeWatch: func(ctx context.Context, opts metav1.ListOptions) (context.Context, metav1.ListOptions, error) {
+			newLabelSelector, newFieldSelector, err := updateSelectorString(ctx, opts.LabelSelector, opts.FieldSelector)
+			if err != nil {
+				return ctx, opts, err
+			}
+			opts.LabelSelector = newLabelSelector
+			opts.FieldSelector = newFieldSelector
+			return ctx, opts, nil
+		},
+	}
+}
+
+type TransformResourceBeforeFunc func(ctx context.Context, resource *unstructured.Unstructured, subresources ...string) (transformed *unstructured.Unstructured, err error)
+type TransformResourceAfterFunc func(ctx context.Context, resource *unstructured.Unstructured, eventType *watch.EventType, subresources ...string) (transformed *unstructured.Unstructured, err error)
+
+func TransformsResource(transformerName string, before TransformResourceBeforeFunc, after TransformResourceAfterFunc) Transformer {
```

Comment:
what is this? Who uses it?

Reply:
Is this NewGenericTransformer? It would be more idiomatic to have:
```golang
type GenericTransformer struct {
  Before: TransformResourceBeforeFunc
  After: TransformResourceAfterFunc
}
```
and then methods for the client methods.